pkg/ingester/transfer.go

```diff
@@ -446,32 +446,61 @@ func (i *Ingester) transferOut(ctx context.Context) error {
 }
 
 func (i *Ingester) v2TransferOut(ctx context.Context) error {
-	if len(i.TSDBState.dbs) == 0 {
-		level.Info(util.Logger).Log("msg", "nothing to transfer")
+	// Skip TSDB transfer if there are no DBs
+	i.userStatesMtx.RLock()
+	skip := len(i.TSDBState.dbs) == 0
+	i.userStatesMtx.RUnlock()
+
+	if skip {
+		level.Info(util.Logger).Log("msg", "the ingester has nothing to transfer")
 		return nil
 	}
 
-	// Close all user databases
-	wg := &sync.WaitGroup{}
-	// Only perform a shutdown once
-	once.Do(func() {
+	// This transfer function may be called multiple times in case of error,
+	// until the max number of retries is reached. For this reason, we run
+	// some initialization only once.
+	i.TSDBState.transferOnce.Do(func() {
+		// In order to transfer TSDB WAL without closing the TSDB itself - which is a
+		// pre-requisite to continue serving read requests while transferring - we need
+		// to make sure no more series will be written to the TSDB. For this reason, we
+		// wait until all in-flight write requests have been completed. No new write
+		// requests will be accepted because the "stopped" flag has already been set.
+		level.Info(util.Logger).Log("msg", "waiting for in-flight write requests to complete")
+
+		waitCtx, waitCancel := context.WithTimeout(ctx, 10*time.Second)
+		defer waitCancel()
+
+		if !util.WaitGroup(waitCtx, &i.TSDBState.inflightWriteReqs) {
```

Comment:
Now if `WaitGroup` returns false, it may indicate various things. If `ctx.Err()` (original context) returns non-nil value here, transfer should not continue.

Reply:
For the records, we discuss it though Slack and we came to the conclusion that we shouldn't use the parent context cause we don't want to interrupt while waiting for in-flight requests to complete if the parent context is cancelled, given this logic run only once.

@pstibrany What's about now?

Reply:
LGTM.